File: Minho/0831_2.cpp
#include <algorithm>
#include <iostream>
using namespace std;

int main() {
    int N, M;
    cin >> N >> M;
    int blue_ray[100001];

    long long sum = 0;
    long long left = -1;

    for (int i = 0; i < N; i++) {
        cin >> blue_ray[i];
        sum += blue_ray[i];
        if (left < blue_ray[i])
            left = blue_ray[i];
    }

    long long right = sum;

    while (left <= right) {
        long long cnt = 0;
        long long middle = (left + right) / 2;
        long long temp_sum = 0;

        for (int i = 0; i < N; i++) {
            if (temp_sum + blue_ray[i] > middle) {
                temp_sum = 0;
                cnt++;
            }
            temp_sum += blue_ray[i];
        }

        if (temp_sum != 0)
            cnt++;

        if (cnt <= M)
            right = middle - 1;
        else
            left = middle + 1;
    }

    cout << left;
}

File: subin/0831/0831_2.java
package fc0831;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class guitarLesson {
    public static int N,M,min,max;
    public static int[] lectureArray;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        String[] input=br.readLine().split(" ");
        N = Integer.parseInt(input[0]);
        M = Integer.parseInt(input[1]);
        lectureArray = new int[N];

        StringTokenizer st=new StringTokenizer(br.readLine()," ");
        for (int i = 0; i < N; i++) {
            lectureArray[i]= Integer.parseInt(st.nextToken());
            min=Math.max(min,lectureArray[i]);
            max+=lectureArray[i];
        }
        System.out.println(binarySearch());

    }
    public static int binarySearch(){
        int left=min;
        int right=max;

        while (left<=right){
            int mid=(left+right)/2;
            int sum=0;
            int cnt=0;
            for (int lecture:lectureArray) {
                //설정한 값보다 합이 크다면 블루레이 쪼개야함
                if(sum+lecture>mid){
                    sum=0;
                    cnt++;
                }
                sum+=lecture;
            }
            //마지막 블루레이 개수 세주기
            if(sum!=0){
                cnt++;
            }
            if(cnt>M){
                left=mid+1;
            }else {
                right=mid-1;
            }
        }
        return left;
    }
}
